rocksdb/sst_file_iterator.go

```diff
@@ -0,0 +1,196 @@
+package rocksdb
+
+import (
+	"os"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	ErrChecksumMismatch    = errors.New("Checksum mismatch")
+	ErrMagicNumberMismatch = errors.New("Magic number mismatch")
+	errEnd                 = errors.New("reach end of block")
+)
+
+type SstFileIterator struct {
+	f              *os.File
+	indexBlockIter *blockIterator
+	dataBlockIter  *blockIterator
+	dataBuf        []byte
+	invalid        bool
+	err            error
+	checksumType   ChecksumType
+}
+
+func NewSstFileIterator(f *os.File) *SstFileIterator {
+	return &SstFileIterator{
+		f:             f,
+		dataBlockIter: new(blockIterator),
+	}
+}
+
+func (it *SstFileIterator) SeekToFirst() {
+	if err := it.loadIndexBlock(); err != nil {
+		it.setErr(err)
+		return
+	}
+	if err := it.loadNextDataBlk(); err != nil {
+		it.setErr(err)
+		return
+	}
+	it.Next()
+}
+
+func (it *SstFileIterator) Next() {
+	if it.dataBlockIter.end() {
+		if err := it.loadNextDataBlk(); err != nil {
+			it.setErr(err)
+			return
+		}
+	}
+
+	it.dataBlockIter.Next()
+}
+
+func (it *SstFileIterator) Key() InternalKey {
+	var ikey InternalKey
+	ikey.Decode(it.dataBlockIter.Key())
+	return ikey
+}
+
+func (it *SstFileIterator) Value() []byte {
+	return it.dataBlockIter.Value()
+}
+
+func (it *SstFileIterator) Valid() bool {
+	return !it.invalid
+}
+
+func (it *SstFileIterator) Err() error {
+	return it.err
+}
+
+func (it *SstFileIterator) loadNextDataBlk() error {
+	var err error
+
+	if it.indexBlockIter.end() {
+		return errEnd
+	}
+
+	it.indexBlockIter.Next()
+	var handle blockHandle
+	handle.Decode(it.indexBlockIter.Value())
+
+	it.changeBufSize(handle.Size + blockTrailerSize)
+	if _, err = it.f.ReadAt(it.dataBuf, int64(handle.Offset)); err != nil {
+		return err
+	}
+	if it.dataBuf, err = it.decompressBlock(it.dataBuf); err != nil {
+		return err
+	}
+	it.dataBlockIter.Reset(it.dataBuf)
+
+	return nil
+}
+
+func (it *SstFileIterator) changeBufSize(sz uint64) {
+	if uint64(len(it.dataBuf)) < sz {
```

Comment:
Use `cap` instead of `len` can save some memory allocation.

Reply:
In this case `len(buf)` always equals to `cap(buf)`.